### PR TITLE
feat(client): ts-migrate WithInstantSearch.js

### PR DESCRIPTION
**Pull request by awu43**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Refers #42256

Converted `InstantSearchRoot` to functional component.

### CODE REVIEW

**Review by ShaunSHamilton — COMMENTED** [review]
Hey there. Thanks again, for another contribution.

I am just double checking, but do you follow the suggestion on the issue for how to rename files and keep their history?:
> Keeping Git History
Sometimes changing the file from <filename>.js -> <filename>.tsx? causes the original file to be deleted, and a new one created, and other times the filename just changes - in terms of Git. Ideally, we want the file history to be preserved.

>The best bet at achieving this is by:
>1) Rename the file
> 2) Commit with the flag --no-verify to prevent husky from complaining about the errors
> 3) Refactor to TS for migration, in a separate commit.

> _Note: Editors like VSCode are still likely to show you the file has been deleted and a new one created. If you use the CLI to git add ., then VSCode will show the file as renamed in stage_

**Comment by awu43** [comment]
> I am just double checking, but do you follow the suggestion on the issue for how to rename files and keep their history?

Yes, the first commit is for renaming files only with no content changes. What's strange is that Git correctly identifies the [renaming commit](https://github.com/freeCodeCamp/freeCodeCamp/pull/42923/commits/b4016bc62954b569657d28d31f5a5357fb75a2d7) and the [migrating commit](https://github.com/freeCodeCamp/freeCodeCamp/pull/42923/commits/431605a2176c1b8e3e6827995d51a366cad04d7b) individually, but not together in [Files changed](https://github.com/freeCodeCamp/freeCodeCamp/pull/42923/files).

**Comment by ShaunSHamilton** [comment]
> > I am just double checking, but do you follow the suggestion on the issue for how to rename files and keep their history?
> 
> Yes, the first commit is for renaming files only with no content changes. What's strange is that Git correctly identifies the [renaming commit](https://github.com/freeCodeCamp/freeCodeCamp/pull/42923/commits/b4016bc62954b569657d28d31f5a5357fb75a2d7) and the [migrating commit](https://github.com/freeCodeCamp/freeCodeCamp/pull/42923/commits/431605a2176c1b8e3e6827995d51a366cad04d7b) individually, but not together in [Files changed](https://github.com/freeCodeCamp/freeCodeCamp/pull/42923/files).

Thanks, for confirming. It is just odd why it sometimes works.

**Comment by ojeytonwilliams** [comment]
It is a bit odd.  For what it's worth, Git Graph is still very readable, as is the last commit.

File: client/src/components/search/with-instant-search.tsx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+import { createSelector } from 'reselect';
+
+const { algoliaAppId, algoliaAPIKey } = envData;
+
+const DEBOUNCE_TIME = 100;
+
+// If a key is missing, searches will fail, but the client will still render.
+const searchClient =
+  algoliaAppId && algoliaAPIKey
```
Comment (ojeytonwilliams) [inline client/src/components/search/with-instant-search.tsx:29]: So, in this case, we only need these to be falsy, not explicitly `null`.  To fix this, could you change config/read-env.js to return `''` for these two properties where it currently returns `null`?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I just want to address the algolia id/key issue first, so that we can get past the linting step.

File: client/src/components/search/with-instant-search.tsx (hunk 1 above)
Comment (awu43) [reply client/src/components/search/with-instant-search.tsx:29]: It's been done.

File: client/src/components/search/with-instant-search.tsx [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+      } else if (query) {
+        updateSearchQuery('');
+      }
+    }
+  });
+
+  const debouncedSetState = useRef<ReturnType<typeof setTimeout>>(
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    setTimeout(() => {})
+  );
```
Comment (ojeytonwilliams) [inline client/src/components/search/with-instant-search.tsx:132]:
```suggestion
  const debouncedSetState = useRef<number>();
```
This is a little more concise and avoids the confusion about which `setTimeout` (browser or node) this is.

File: client/src/components/search/with-instant-search.tsx [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  function updateBrowserHistory(query: string): void {
+    if (isSearchPage()) {
+      clearTimeout(debouncedSetState.current);
+
+      debouncedSetState.current = setTimeout(() => {
```
Comment (ojeytonwilliams) [inline client/src/components/search/with-instant-search.tsx:137]:
```suggestion
      debouncedSetState.current = window.setTimeout(() => {
```
If we use `number` above, we have to be explicit.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Thanks for fixing the conflicts, @awu43

This PR does look good, but one general observation: it's difficult to properly review (and potentially debug) PRs that do multiple things at once.  In this case, we can't simply look at the diff to see which changes are just part of the the TS migration and which are caused by the change to use functional components.

That aside, there's one nitpick below:

**Comment by awu43** [comment]
Good to know about the component types, I'll keep it in mind.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

Thanks @awu43 this looks solid.

**Review by ShaunSHamilton — APPROVED** [review]
This LGTM 👍

**Comment by ShaunSHamilton** [comment]
Thank you, for your contribution to the page! 👍
We are happy to accept these changes ,and look forward to future contributions. 🎉

**Pull request merged**